### PR TITLE
View source jar files as virtual docs

### PR DESCRIPTION
**Pull request by Arthurm1**
Currently when the client wants to browse files in source jars, the files are extracted and saved to `.metals/readonly/dependencies`
e.g.
`file:///workspacePath/.metals/readonly/dependencies/src.zip/java.base/java/nio/file/Path.java`
`file:///workspacePath/.metals/readonly/dependencies/trees_2.12-4.4.28-sources.jar/scala/meta/io/AbsolutePath.scala`

Current downsides:  The extracted source files aren't read only and they have to be written and cleared up.
Virtual doc downsides: These files won't be re-opened when VSCode is shutdown and re-opened

This PR means jar/jdk sources are read directly and passed back to the client as text.  This method uses the original URIs...
e.g.
`jar:file:///usr/lib/jvm/adoptopenjdk-11-hotspot-amd64/src.zip!/java.base/java/nio/file/Path.java`
`jar:file:///userPath/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scalameta/trees_2.12/4.4.28/trees_2.12-4.4.28-sources.jar!/scala/meta/io/AbsolutePath.scala`

This setting can be set per client in `MetalsServerConfig#isVirtualDocumentSupported`

There are a number of places that URIs are passed around as text and transformed into `AbsolutePath`.  That's fine for  `/somepath` but not `jar:file:///jarpath!package/file.java` so these transforms have to go through URI.  I may have missed some of these as I don't know all the code paths from VSCode->Metals.

Also - I'm working around https://github.com/coursier/coursier/issues/2186 (although if this issue is correct then I don't need to work around) but I think the workaround in `MetalsEnrichments#XtensionString#toAbsolutePath` can be left as is.

This PR depends on #3107

### CODE REVIEW

**Comment by ckipp01** [comment]
> This setting can be set per client in MetalsServerConfig#isVirtualDocumentSupported

I'd actually recommend only supporting and documenting this via `InitializationOptions`. This sort of seems like the best place to support it (I see you already do), so I'm not sure if there is any benefit to also adding it to `MetalsServerConfig`. That was sort of the very initial place settings used to be added before we started using `ClientExperimentalCapablities` or `InitializationOptions`. Now everything that is client specific should be added via `InitialziationOptions`.

File: metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala:341]: This makes sure that we don't save the file to disk, but I can understand how the file is being sent back to the user. Do we send the path with the jar and VS Code is responsible for that?

**Review by tgodzik — CHANGES_REQUESTED** [review]
I left some minor comments, but looks good. Could you take a look at the failing tests? Some of them are flaky, but it seems that some of the failures are real.

File: metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala
Comment (Arthurm1) [reply metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala:341]: Yes - the wonder of Virtual documents.  Metals tells VSCode (or any client) to open a jar URI e.g. `jar:file://path/jarname.jar!source/source.java` instead of the URI of the read-only file that we've extracted from the jar and saved on the file system.  VSCode intercepts any URI with the `jar` scheme and asks the `MetalsFileProvider` to provide the contents of that document.  `MetalsFileProvider` sends a `DecodeFile` command to Metals with this same URI. Metals hands this off to the `FileDecoderProvider` which extracts the file's contents from the jar and sends it back to VSCode.  No persistence to the filesystem occurs.  It's quite a nice way (I think) of handling any data that we don't have saved on the file system but refer to by URI.  Then you can use standard jar style URIs to navigate round the source jars or still create read-only files and still navigate round source jars using `file` scheme URIs.  If all clients supported this then Metals could dump the readonly folder entirely (I think)

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala:341]: Thanks for explanation!

File: metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala
Comment (dos65) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:444]: There functions are correct in 99% percents but I believe it might cause some exceptionally rare issues that would be hard to investigate.

There is non zero possibility that there might a be a dependency jar without a version that ends with `-sources` in build - some local unmanaged jar.

The same for `src.zip` - probably there might be some `src.zip` that aren't about Jdk at all. Also, I'm not sure if all jdk/os distros put sources in `src.zip`

It would be more reliable always check it if jar is a source in `BuildTargets` as it's anyway quite fast.

**Comment by Arthurm1** [comment]
I think this is finished now.  The Windows test failure is a timeout in `DebugDiscoverySuite`

Paired with https://github.com/scalameta/metals-vscode/pull/695

**Review by kpodsiad — APPROVED** [review]

**Review by tgodzik — CHANGES_REQUESTED** [review]
There seem to be some issues with Tree View, which I am taking a look at here:
https://github.com/scalameta/metals/compare/main...tgodzik:source_jars_virtual_documents?expand=1

I will update the PR with the fixes later on.

**Comment by Arthurm1** [comment]
I'm off on holiday for a week so I can look when I get back - I don't think there's a rush for this feature as it doesn't add any functionality

**Comment by tgodzik** [comment]
> I'm off on holiday for a week so I can look when I get back - I don't think there's a rush for this feature as it doesn't add any functionality

I will take a quick look then and maybe focus on some missing features in Java support instead.

**Comment by Arthurm1** [comment]
There's an issue with calculating interactive semanticdb files on jar classes now that they are virtual documents rather than files in the readonly dir.

For Java, I'm not sure it's possible without saving to file first as I don't think you can call javac on a single file in a jar.  For Scala files - I haven't looked into yet.

One thing I didn't understand though: In `InteractiveSemanticdbs` there is this code...
```scala
 def shouldTryCalculateInteractiveSemanticdb = {
      source.isLocalFileSystem(workspace) && (
        unsavedContents.isDefined ||
          source.isInReadonlyDirectory(workspace) || // dependencies
          source.isSbt || // sbt files
          source.isWorksheet || // worksheets
          doesNotBelongToBuildTarget // standalone files
      )
    }
```
It seems as though semanticdb info would never be calculated on dependencies unless they've been unzipped into the readonly directory.  Is interactive semanticdb info only required if the user has the file open?  e.g. Metals doesn't need semanticdb info on `java.time.LocalDate` unless it's being viewed?  It wouldn't need the info if the user is just hovering over a `LocalDate` reference or using `Goto definition` on it?

**Comment by dos65** [comment]
@Arthurm1 

> For Java, I'm not sure it's possible without saving to file first as I don't think you can call javac on a single file in a jar. For Scala files - I haven't looked into yet.

You right. For Java it will be required to dump into a file but I don't think that it's a problem.
For Scala there is no need in doing this as we directly pass virtual source to Presentation Compiler.

> It seems as though semanticdb info would never be calculated on dependencies unless they've been unzipped into the readonly directory. Is interactive semanticdb info only required if the user has the file open? e.g. Metals doesn't need semanticdb info on java.time.LocalDate unless it's being viewed? It wouldn't need the info if the user is just hovering over a LocalDate reference or using Goto definition on it?

Yep, for dependencies semanticdb is required only if this file is opened in the editor.
Actually, it's the same for other types of source files in most cases except of `Find References`/`Rename` features that work differently or shouldn't work for dependency sources.

For example, if you use `java.time.LocalDate` in source `A.scala` then there will be occurrence of `java.time.LocalDate`  in `A.scala.semanticdb` and that's is enough information to perform goto-definition from `A.scala` to `j/t/LocalDate.java`.

**Comment by olafurpg** [comment]
This is exciting! I just want to leave a short note that VS Code’s FileSystemProvider is preferable over TextDocumentContentProvider because it gives you breadcrumbs in the UI and the ability to list files in a directory (making it easy to explore the entire jar file instead of only a single file). Implementing the file system API may require slightly differently shaped messages in the JSON-RPC layer.

**Comment by Arthurm1** [comment]
@olafurpg I did have a brief look at it but it needs this PR done first as Metals doesn't understand `jar:` URIs everywhere yet.

The way VSCode handles FileSystemProvider is as an additional workspace.  Metals doesn't understand multiple workspaces yet but it's not that difficult to handle this specific feature.  It does then mean breadcrumbs work, but I'm not sure there's any benefit over the current MetalsTreeView which already allows the user to explore the jar.

Maybe it would mean we could drop the `libraries` section of the the TreeView and make the support easier across non-vscode clients, although VIM does support the treeview.  I can't see why we'd keep both FileSystemProvider and TreeView:Libraries.

**Comment by olafurpg** [comment]
With FileSystemProvider you still need to implement your own tree views (vscode doesn’t provide one automatically), so the libraries view can stay unchanged. The main selling point is the breadcrumbs, which is a big deal IMO because it makes it much easier to see what file you have open (especially if the URI has a human readable way to identify the library name and version). The vscode proposed api has new methods to integrate with the native file tree so there may be other benefits in the future.

**Comment by olafurpg** [comment]
I’ve been working with the vscode file system APIs lately and just wanted to share these findings. No need to act on it, having virtual documents alone would be a nice huge improvement already :)

**Comment by Arthurm1** [comment]
That makes sense.  I do think breadcrumbs are handy - I just thought I was missing something else.

**Comment by Arthurm1** [comment]
I'm struggling to get this working in all scenarios.  I can either get `ImplementationLspSuite` to pass or `JavaDefinitionVirtualDocSuite` but not both.

The difference in code to get them working is `InteractiveSemanticdbs`.  I either have the current...
```scala
    def shouldTryCalculateInteractiveSemanticdb = {
      source.isLocalFileSystem(workspace) && (
        unsavedContents.isDefined ||
          source.isInReadonlyDirectory(workspace) || // dependencies
          source.isSbt || // sbt files
          source.isWorksheet || // worksheets
          doesNotBelongToBuildTarget // standalone files
      )
    }
```

or the new...
```scala
    def shouldTryCalculateInteractiveSemanticdb = {
      source.isLocalFileSystem(workspace) && (
        unsavedContents.isDefined ||
          source.isInReadonlyDirectory(workspace) || // dependencies
          source.isSbt || // sbt files
          source.isWorksheet || // worksheets
          doesNotBelongToBuildTarget // standalone files
      ) || source.isJarFileSystem // VIRTUAL DOC DEPENDENCIES
    }
```
The last line is the difference.  It's needed to get the [JavaDefinitionSuite#jdk-String-patch-module](https://github.com/scalameta/metals/blob/bc2fa7a60be8369e073ef0b9aab363d09e658d31/tests/unit/src/test/scala/tests/JavaDefinitionSuite.scala#L36) test to pass, as `java.lang.String` file must have semanticdb generated for it.  But then semanticdb gets generated for `java.lang.Exception` in the [ImplementationLspSuite#java-classes](https://github.com/scalameta/metals/blob/bc2fa7a60be8369e073ef0b9aab363d09e658d31/tests/unit/src/test/scala/tests/ImplementationLspSuite.scala#L253) test which causes `ImplementationProvider` to fail to find everything.

I can see 2 ways of fixing this but I'm not sure the correct route.  Currently it works because in the above `InteractiveSemanticdbs` code, it will only generate semanticdb info if the file exists in the readonly area.  For `JavaDefinitionSuite#jdk-String-patch-module`, the test code calls `MetalsLanguageServer#workspaceSymbol` which saves `java.lang.String` to readonly area so it has semanticdb run on it.  In `ImplementationLspSuite#java-classes` the `java.lang.Exception` class is not saved to disk so no semanticdb is generated.

With virtual docs, this info about whether a file has or hasn't been saved in the readonly area is lost.  Every dependency uri is `jar:file//XXXX.jar` and is never saved.  So there is no info at the point in `InteractiveSemanticdbs` to decide whether we should or shouldn't generate semanticdb info. So we always have to on a `.scala` or `.java` file.

You'd think that this wouldn't be an issue - if we always generate it then the worst thing that can happen is a drop in performance but in [ImplementationProvider#implementations](https://github.com/scalameta/metals/blob/bc2fa7a60be8369e073ef0b9aab363d09e658d31/metals/src/main/scala/scala/meta/internal/implementation/ImplementationProvider.scala#L131) we can see different paths taken depending on if semanticdb info exists.  Either it uses `globalTable#globalContextFor` if there is no semantidb info or `InheritanceContext#fromDefinitions` if there is.  But the issue is that `InheritanceContext#fromDefinitions` isn't aware of all the other inheritance info, e.g. that `RuntimeException` inherits from `Exception`, so doesn't return all of the implementations - only some of them.

So should `ImplementationProvider#implementations` be changed to check `InheritanceContext` _and_ `globalTable`?  Or should `Semanticdbs#textDocument` be changed to take more information than just a `path` e.g. `generateIfMissing: Boolean`.  Or should `InheritanceContext` have had the info on `RuntimeException` and how would it get that info? 

I just don't know which is supposed to be correct as I'm not sure what info should be stored where.

**Comment by tgodzik** [comment]
@Arthurm1 I think this might actually be a bug, I did a quick fix for your use case here:

https://github.com/tgodzik/metals/commit/0d48e8fe27ec6bc31f40ecc93d75a26a749e1d0f

But we should not calculate semanticdb for ImplementationProvider, I will check if this is an issue in main also.

**Comment by tgodzik** [comment]
Ok, interestingly enough, this can also happen in the current main if the user goes to definition on an Exception. I will look into that.

I noticed that there is an exception popping up:
```
2022.01.18 14:10:01 ERROR text document: file:///Users/runner/work/metals/metals/tests/unit/target/e2e/implementation/java-classes/jar%253Afile%253A%252F%252F%252FUsers%252Frunner%252FLibrary%252FCaches%252FCoursier%252Fv1%252Fhttps%252Frepo1.maven.org%252Fmaven2%252Forg%252Fscala-lang%252Fscala-library%252F2.12.15%252Fscala-library-2.12.15-sources.jar%2521%252Fscala%252Fpackage.scala
java.nio.file.NoSuchFileException: /Users/runner/work/metals/metals/tests/unit/target/e2e/implementation/java-classes/jar%3Afile%3A%2F%2F%2FUsers%2Frunner%2FLibrary%2FCaches%2FCoursier%2Fv1%2Fhttps%2Frepo1.maven.org%2Fmaven2%2Forg%2Fscala-lang%2Fscala-library%2F2.12.15%2Fscala-library-2.12.15-sources.jar%21%2Fscala%2Fpackage.scala
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)
	at java.nio.file.Files.newByteChannel(Files.java:371)
	at java.nio.file.Files.newByteChannel(Files.java:422)
	at java.nio.file.Files.readAllBytes(Files.java:3206)
	at scala.meta.internal.io.PlatformFileIO$.slurp(PlatformFileIO.scala:45)
	at scala.meta.internal.io.FileIO$.slurp(FileIO.scala:24)
	at scala.meta.internal.metals.InteractiveSemanticdbs.$anonfun$textDocument$2(InteractiveSemanticdbs.scala:90)
	at scala.Option.getOrElse(Option.scala:189)
...
```
this might be related.

**Comment by Arthurm1** [comment]
There's a couple of those where `TextDocument#uri = workspace.resolve(path)` is called on a jar.

`uri` is a `String` so pretty much anything can be put in there.  Sometimes a real `jar:file` uri comes in and then you get an exception as we try to resolve that against the workspace.  Most of the time `path` is a relative location in the workspace. e.g. `.metals/readonly/someunzippedjarfile.scala`.

I didn't realise this was happening on main.

Ideally I'd like `TextDocument#uri` to always be a proper uri e.g. 'jar:file///xxx' or 'file:/xxx' or 'metalsdecode:/XXX' but I don't know if this affects other code

**Comment by tgodzik** [comment]
> There's a couple of those where `TextDocument#uri = workspace.resolve(path)` is called on a jar.
> 
> `uri` is a `String` so pretty much anything can be put in there. Sometimes a real `jar:file` uri comes in and then you get an exception as we try to resolve that against the workspace. Most of the time `path` is a relative location in the workspace. e.g. `.metals/readonly/someunzippedjarfile.scala`.
> 

I think that if we could check if it's a jar and then use the relative position in the jar. Or something along these lines?

> I didn't realise this was happening on main.

Actually I checked again and that seems to not be an issue as we don't direct to readonly directory in ImplementationProvider, it's always the jar, which is why it's popping up here. But the change I suggested here should be ok.

**Comment by Arthurm1** [comment]
Not sure why the Windows test fails - flaky?

I think this is complete.  There's a number of tests that save jar classes to the `readonly` area.  I've subsclassed all of those so they test virtual docs and saveToDisk.  Hopefully I've caught all of them.

I've repurposed `readonlySources` in TestingServer.  It wasn't being used, so it now records jar names.  The issue with `TestingServer` is that it can easily find jar classes when they're saved to disk as they're all in the `dependencies` area.  When they're just accessed virtually they aren't saved anywhere so this `TrieMap` records that.

Jars are a pain to work with as the FileSystem has to be opened to just define an `AbsolutePath` let alone use methods on it.  Because of that I've left the FileSystem open for any jar that's defined using `URI.toAbsolutePath`.  Otherwise every `AbsolutePath` method would have to be wrapped in `PlatformFileIO.newFileSystem`.  It looks like Jars are already left open in `JarTopLevels#getTopLevels`.

In `ClientConfigurationAdapter` I've passed a `URI` when it should be a path but then there's no such thing for jars is there - it's always a URI?  I'm not sure why we don't use `InitializeRequestArgumentsPathFormat.URI` as the format but passing a jar URI works in both cases.

I've tested References, Goto, TreeView sync, debug stacktrace navigation, analyse stacktrace navigation.  Is there anything else?

**Comment by tgodzik** [comment]
> Not sure why the Windows test fails - flaky?

I am rerunning them, but it's the first time I've seen them running out of memory :thinking: 

> I've tested References, Goto, TreeView sync, debug stacktrace navigation, analyse stacktrace navigation. Is there anything else?

Nothing currently coming to mind. I will have a try before merging it, but otherwise we can test it out a bit more on the newest snapshots, so fortunately we can iron it out before the next release (which this could be a main feature of :rocket: )

**Comment by Arthurm1** [comment]
@tgodzik Cool.  Wouldn't mind if you also merged the `Display Build Info` functionality 😉

**Comment by tgodzik** [comment]
> Cool. Wouldn't mind if you also merged the Display Build Info functionality

Sure! I can look at it first, looks shorter to review :S

**Comment by tgodzik** [comment]
The windows failures look legit:
```
java.lang.Exception: Obtained [file:///D:/a/metals/metals/tests/unit/target/e2e/debug-step-save-to-disk/step-out/a/src/main/scala/Main.scala, 5], expected [D:\a\metals\metals\tests\unit\target\e2e\debug-step-save-to-disk\step-out\a\src\main\scala\Main.scala, 5]
```

File: metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:550]:
```suggestion
                create = false
```
? 

Not 100% sure what create does, but we sue it in a lot of places. I asume it means it will not throw an exception if the jar system is already created?

File: tests/slow/src/test/scala/tests/feature/WorksheetCrossLspSuite.scala
Comment (tgodzik) [inline tests/slow/src/test/scala/tests/feature/WorksheetCrossLspSuite.scala:13]: We can run `virtual-docs` as default and let's just add a single test suite (could be case really) that runs with `save-to-disk`. I don't think we need to replicate it for every Scala versions and does take a while to run it. We can even remove the `2.11` suite I think, since we did depracte it.

File: tests/slow/src/test/scala/tests/sbt/SbtStepDapSuite.scala
Comment (tgodzik) [inline tests/slow/src/test/scala/tests/sbt/SbtStepDapSuite.scala:17]: I would avoid running the tests double and just run a single test case with changed useVirtualDocuments option. These tests do take a while and I think is is also partly the reason why Windows tests run out of memory at some point.

File: tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala
Comment (tgodzik) [inline tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala:17]: We could make `initializationOptions` a var and change `useVirtualDocuments` for a single case instead.

File: tests/unit/src/test/scala/tests/RenameLspSuite.scala
Comment (tgodzik) [inline tests/unit/src/test/scala/tests/RenameLspSuite.scala:5]: RenameLsp suite take a lot of time, so this would be really bad for out CI to run it twice.

**Review by tgodzik — CHANGES_REQUESTED** [review]
Thanks for the great work! This is a lot of changes, far more than I expected it to need.

Please avoid running the suites double as it may add a lot of running time to our CI and the windows jobs are already at an edge (I think the CI failed previously because of that). Could we instead have a single case everywhere that checks the specific logic when using readonly directory? I would assume the virtual docs to be the default for all the other cases.

We can do it 2 ways:
- make init options a var and replace it for a single test (find one that was failing for that suite when changed to virtual docs)
- have a separate test case for each suite that runs only when `.readonly` files option is ticket and everything else in other case

I need to rewrite some of the suites and move a lot the logic testing to some unit tests while leaving only real integration tests there.

Again, thanks for the great work and it already looks amazing. Sorry for the additional comments. If you unable to find time to do it, I can do it next week.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala
Comment (Arthurm1) [reply metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala:550]: I think `create` is there if you want to create a jar from scratch.  The jars here should all exist.  Also - that param isn't available on the method I'm calling as I'm passing a URI, not an AbsolutePath

File: metals/src/main/scala/scala/meta/internal/metals/JavaInteractiveSemanticdb.scala
Comment (Arthurm1) [reply metals/src/main/scala/scala/meta/internal/metals/JavaInteractiveSemanticdb.scala:42]: You mean the `workDir`?  I.e. wrap everything in a `try..finally`?

**Comment by Arthurm1** [comment]
@tgodzik I'm not sure exactly what you want with altering the tests.

You mean just preserve 1 test within each suite to test the `save-to-disk` version?

I'll have to look through all the tests to find out which individual test actually writes to disk - before I just checked if the testsuite wrote to disk.

I get your point about not wanting to increase the testing time - I had to double up on 16 testsuites which is a lot.

**Comment by tgodzik** [comment]
> You mean just preserve 1 test within each suite to test the save-to-disk version?

That's what I meant, I think that should be enough to test out all the issues that might pop up. It's not perfect for sure, but until we manage to get the test more effective it will have to do.

**Comment by tgodzik** [comment]
> I'll have to look through all the tests to find out which individual test actually writes to disk - before I just checked if the testsuite wrote to disk.

I will take a look at it.

**Comment by tgodzik** [comment]
I added a PR here: https://github.com/Arthurm1/metals/pull/1

That should work and not add too much testing time.

**Comment by Arthurm1** [comment]
@tgodzik I've tried out your testsuite fix but it's a bit temperamental.

I think this is because `newServer` adjusts the `initializationOptions` depending on the value of `useVirtualDocs` but `newServer` is called in `beforeEach` and the `virtualDocs = true/false` is set in the test body which would be _after_ the `newServer` has been setup.

The only ways I can think of a way of influencing the `newServer` call are...
1. Go back to subclassing the suites which we're trying to avoid to reduce test times
2. Shift `newServer` into the `test` body - but then it would no longer be aware of the test name
3. Add a `VirtualDoc` test tag.  Which is what I've done here.  What do you think?  I don't really like it - it feels wrong - but it could be temporary if we eventually drop `savetodisk`.


Also - I see you've added `test("definition".only` into `BaseWorksheetLspSuite` but it fails on CI with the following error message that I don't understand...
```
[error] ==> X tests.worksheets.WorksheetLspSuite.definition-readonly  0.003s munit.FailException: /home/runner/work/metals/metals/tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala:494 'Only' tag is not allowed when `isCI=true`
```

File: tests/unit/src/main/scala/tests/BaseLspSuite.scala
Comment (tgodzik) [inline tests/unit/src/main/scala/tests/BaseLspSuite.scala:103]: We could add an additional option to `newServer` 
```suggestion
  def newServer(workspaceName: String, useVirtualDocs): Unit = {
```
and check in beforeEach if the tag is set, which would probably be better than the var I proposed.

File: tests/unit/src/main/scala/tests/BaseLspSuite.scala
Comment (tgodzik) [inline tests/unit/src/main/scala/tests/BaseLspSuite.scala:54]: I think the tag approach is ok with a small modification for removing the var.

File: tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala
Comment (tgodzik) [inline tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala:494]: This is my bad, `.only` is one of the default tags that will run only the test case tagged with it and it breaks on CI in case someone (such as me here) forgets to remove it :sweat:

**Review by tgodzik — CHANGES_REQUESTED** [review]
> I've tried out your testsuite fix but it's a bit temperamental.

>I think this is because newServer adjusts the initializationOptions depending on the value of useVirtualDocs but newServer is called in beforeEach and the virtualDocs = true/false is set in the test body which would be after the newServer has been setup.

Ach you are right, I somehow thought it was working, but maybe I tested it out too quickly. Your approach with the tags is ok I think.

As for the `.only` issue that's also my bad, added an explanation in the comment. Sorry!

File: tests/unit/src/main/scala/tests/BaseLspSuite.scala
Comment (Arthurm1) [reply tests/unit/src/main/scala/tests/BaseLspSuite.scala:54]: Do you mean removing the `val virtualDocTag`?  If so - where would you put it - in a companion object?  Or do you mean remove the `val useVirtualsDoc` var?

File: tests/unit/src/main/scala/tests/BaseLspSuite.scala
Comment (Arthurm1) [reply tests/unit/src/main/scala/tests/BaseLspSuite.scala:103]: The var is still needed as the tests themselves need to know whether they are using virtualdocs, so need to be able to call `BaseLspSuite#useVirtualDocuments`.  And it can't just interrogate `initializationOptions` as this doesn't contain the virtualdoc info as the `initializationOptions` are added to in `newServer`.  We could add the additional option to `newServer` but it would just be duplicating the `BaseLspSuite#useVirtualDocs` var

Comment (tgodzik) [reply tests/unit/src/main/scala/tests/BaseLspSuite.scala:103]: Ok, makes sense!

File: tests/unit/src/main/scala/tests/BaseLspSuite.scala
Comment (tgodzik) [reply tests/unit/src/main/scala/tests/BaseLspSuite.scala:54]: I meant the var, but if we still need it then it's ok.

**Comment by Arthurm1** [comment]
@tgodzik I think everything is addressed now and ready to merge unless you wanted me to wrap the `JavaInteractiveSemanticdb#textDocument` in a `Try` but the `javac` command itself is wrapped in a `Future` and the error handled.

File: tests/unit/src/test/scala/tests/WorkspaceSymbolLspSuite.scala
Comment (tgodzik) [inline tests/unit/src/test/scala/tests/WorkspaceSymbolLspSuite.scala:19]: I wonder if we could make something line `TestDefault` in the `BaseLspSuite` ? So that we don't repeat the init options everywhere.

File: tests/unit/src/test/scala/tests/debug/StepDapSuite.scala
Comment (tgodzik) [inline tests/unit/src/test/scala/tests/debug/StepDapSuite.scala:8]:
```suggestion
      "debug-step",
```

**Review by tgodzik — CHANGES_REQUESTED** [review]
I think we are almost good to merge, I left some minor comments.

**Comment by Arthurm1** [comment]
@tgodzik I think I've addressed everything

I've added `cleanWorkspace` to all the `WorksheetLspSuite` tests.  They were failing for me because I had an old H2 database around in the test area and it kept throwing errors.

**Review by tgodzik — APPROVED** [review]
LGTM!

**Pull request merged**
